### ISSUE TITLE
State with `onchanges`, `watch`, `prereq`, `unlsess`, `onlyif` always return True

### ISSUE THREAD

**Issue by angeloudy** [saltstack/salt#54213]
I think this is a problem introduced when the requisite was first designed.
A requiring state would not run if the required state fails.
A state is considered failure if `ret['result'] = True`.
This means as long as the required state has a return value of `True`, the requiring state will run.

Example 1.

```
test-required-state:
  file.managed:
    - name: /tmp/testfile1
    - contents: 'hello world'
    - unless:
       - ls -d /tmp

test-requiring-state:
  cmd.run:
    - name: 'hello world'
    - require:
       - test-required-state

```

Since the `unless` condition is true, I would expect `test-requiring-state` not run, but the result is otherwise.

```
local:
----------
          ID: test-required-state
    Function: file.managed
        Name: /tmp/testxxx
      Result: True
     Comment: unless condition is true
     Started: 10:46:38.203416
    Duration: 3156.125 ms
     Changes:
----------
          ID: test-requiring-state
    Function: cmd.run
        Name: echo "hello"
      Result: True
     Comment: Command "echo "hello"" run
     Started: 10:46:41.360514
    Duration: 12.33 ms
     Changes:
              ----------
              pid:
                  95466
              retcode:
                  0
              stderr:
              stdout:
                  hello

Summary for local
------------
Succeeded: 2 (changed=1)
Failed:    0
------------
Total states run:     2
Total run time:   3.168 s
```

It's easier to see why from the json output `--out=json`

```
{
    "local": {
        "file_|-test-required-state_|-/tmp/testxxx_|-managed": {
            "result": true,
            "name": "/tmp/testxxx",
            "changes": {},
            "comment": "unless condition is true",
            "skip_watch": true,
            "__sls__": "testreq",
            "__run_num__": 0,
            "start_time": "10:46:07.561408",
            "duration": 16663.112,
            "__id__": "test-required-state"
        },
        "cmd_|-test-requiring-state_|-echo \"hello\"_|-run": {
            "name": "echo \"hello\"",
            "changes": {
                "pid": 95380,
                "retcode": 0,
                "stdout": "hello",
                "stderr": ""
            },
            "result": true,
            "comment": "Command \"echo \"hello\"\" run",
            "__sls__": "testreq",
            "__run_num__": 1,
            "start_time": "10:46:24.225505",
            "duration": 14.885,
            "__id__": "test-requiring-state"
        }
    }
}
```

The `require` requisite looks at the value `result` in the returned dictionary, while `watch, onchanges` check `changes` in the returned dictionary.

The difference is if the required state return `False`, the requiring state will also return `False`, and thus you can see comments like `One or more requisite failed`, while states with `onchanges, watch` will always return `True`.

I think this is the root cause of issue #53379 #52200

**Comment by dmurphy18** [comment]
@angeloudy Apologies, I had filed a large comment explaining how this is working correctly as designed, when I added _Expected Behavior_ and forgot to hit _Comment_ button

Basically, the first state returns Result: True since it operates correctly as per the following:
1. Directory does not exist, hence creates file
2. Directory does exist, hence does not create file.

Either way it the operation of the state executes without error, hence Result: True

root@deb9min:/srv/salt# salt t9 state.sls testb
t9:
----------
          ID: test-required-state
    Function: file.managed
        Name: /tmp/testfile1
      Result: True
     Comment: unless condition is true
     Started: 10:07:34.949051
    Duration: 1079.534 ms
     Changes:   

Summary for t9
------------
Succeeded: 1
Failed:    0
------------
Total states run:     1
Total run time:   1.080 s
root@deb9min:/srv/salt# ls -d /tmp
/tmp
root@deb9min:/srv/salt# echo $?
0
root@deb9min:/srv/salt# ls -l /tmp/testfile1
ls: cannot access '/tmp/testfile1': No such file or directory
root@deb9min:/srv/salt# cat testb.sls 
test-required-state:
  file.managed:
    - name: /tmp/testfile1
    - contents: 'hello world'
    - unless:
       - ls -d /tmp

**Comment by dmurphy18** [comment]
```
root@deb9min:/srv/salt# cat teste.sls
test-required-state:
  file.managed:
    - name: /tmp/testfile1
    - contents: 'hello world'
    - unless:
       - ls -d /tmpnotexist

root@deb9min:/srv/salt# ls -d /tmpnotexist
ls: cannot access '/tmpnotexist': No such file or directory
root@deb9min:/srv/salt# echo $?
2
root@deb9min:/srv/salt# salt t9 state.sls teste
t9:
----------
          ID: test-required-state
    Function: file.managed
        Name: /tmp/testfile1
      Result: True
     Comment: File /tmp/testfile1 updated
     Started: 10:11:41.760550
    Duration: 1079.562 ms
     Changes:   
              ----------
              diff:
                  New file

Summary for t9
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:   1.080 s
root@deb9min:/srv/salt# ls -d /tmpnotexist
ls: cannot access '/tmpnotexist': No such file or directory
root@deb9min:/srv/salt# ls -l /tmp/testfile1
-rw-r--r-- 1 root root 12 Aug 16 10:11 /tmp/testfile1
root@deb9min:/srv/salt# echo $?
0
root@deb9min:/srv/salt# 
```

Hopefully this explanation explains better why this is _Expected Behavior_

Please consider closing this issue, unless I can explain further how this is operating as expected.

Note  the second state in incorrect and should be:
```
test-requiring-state:
  cmd.run:
    - name: |
        echo 'hello world'
    - require:
       - test-required-state
```

Since as stated in the original issue statement would have _cmd.run_ execute *_ /bin/bash/hello _*

**Comment by angeloudy** [comment]
That makes sense.
I understand all of that.
It's just the documentation of `prereq` is confusing.


'''
If the pre-requiring state executes successfully, the pre-required state will then execute. If the pre-requiring state fails, the pre-required state will not execute.
'''

I would expect that if the pre-requiring state does not return any changes, neither the pre-required state nor pre-requiring state would actually execute. But that's not the case.

I used to experience huge outage with `prereq` because `archive.py` couldn't return `changes` when `test=true`

One more question: I noticed that most states in test mode returns `None`, what does it mean?
I also had some problem in the past with the state returning `None` in test mode.
